### PR TITLE
Simplify CMake Flag checkers

### PR DESCRIPTION
**Pull request by DesWurstes**
Now, it's obvious which options run on which configurations.

*Ready!*

### CODE REVIEW

**Comment by kylophone** [comment]
lgtm, ping again after your tiny fix is complete

**Comment by DesWurstes** [comment]
@kylophone @1480c1 Please re-review.

Changes:

Screen Shot 2019-07-15 at 18 44 24
Screen Shot 2019-07-15 at 18 44 34

On macOS/GCC, there's only flag to be tested. I've had to add fake flags to fix handling of multiple flags. That's why I didn't encounter the issue.

File: Source/Lib/Common/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,17 @@
 
 # Libraries Directory CMakeLists.txt
 
+function(test_apply_compiler_flags flags current_flags)
+    foreach(cflag ${${flags}})
+        string(REGEX REPLACE "[^A-Za-z0-9]" "_" cflag_var "${cflag}")
+        set(test_c_flag "C_FLAG${cflag_var}")
+        check_c_compiler_flag(${cflag} "${test_c_flag}")
+        if(${test_c_flag})
+            set(${current_flags} "${${current_flags}} ${cflag}" PARENT_SCOPE)
```
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]: Careful for repeating dupes of current flags and remove brackets from first currently flags, it expands before setting

**Comment by 1480c1** [comment]
> On macOS/GCC, there's only flag to be tested. I've had to add fake flags to fix handling of multiple flags. That's why I didn't encounter the issue.

What do you mean "there's only flag to be tested"? Which flags?

**Comment by DesWurstes** [comment]
None of the `if` branches run after `set(flags_to_test -msse4.1)`, etc... So the list `flags_to_test` has always has one element on macOS.

File: Source/Lib/Common/CMakeLists.txt (hunk 1 above)
Comment (DesWurstes) [reply Source/Lib/Common/CMakeLists.txt]: No, CMake functions treat input variables as pointers. It's impossible to understand/guess without experimenting, there's no intuition about CMake pointers.

File: Source/Lib/Common/ASM_AVX2/CMakeLists.txt [hunk 2]
```diff
@@ -15,35 +15,23 @@ include_directories(${PROJECT_SOURCE_DIR}/Source/API/
...
-        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
+        list(APPEND flags_to_test -static-intel -w)
     endif()
 endif()
 
+if(MSVC)
+    list(APPEND flags_to_test /arch:AVX2)
+endif()
+
+test_apply_compiler_flags(flags_to_test CMAKE_C_FLAGS)
```
Comment (1480c1) [inline Source/Lib/Common/ASM_AVX2/CMakeLists.txt]:
```suggestion
test_apply_compiler_flags(${flags_to_test})
```

File: Source/Lib/Common/ASM_SSE2/CMakeLists.txt [hunk 3]
```diff
@@ -11,23 +11,14 @@ include_directories(${PROJECT_SOURCE_DIR}/Source/API/
...
-    endif()
-endforeach()
-
-if(CMAKE_C_COMPILER_ID STREQUAL "Intel")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
+if(CMAKE_C_COMPILER_ID STREQUAL "Intel" AND NOT WIN32)
+    list(APPEND flags_to_test -static-intel -w)
 endif()
 
+test_apply_compiler_flags(flags_to_test CMAKE_C_FLAGS)
```
Comment (1480c1) [inline Source/Lib/Common/ASM_SSE2/CMakeLists.txt]:
```suggestion
test_apply_compiler_flags(${flags_to_test})
```

File: Source/Lib/Common/ASM_SSE4_1/CMakeLists.txt [hunk 4]
```diff
@@ -14,23 +14,13 @@ include_directories(${PROJECT_SOURCE_DIR}/Source/API/
...
-    endif()
-endforeach()
-
-if(CMAKE_C_COMPILER_ID STREQUAL "Intel")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
+if(CMAKE_C_COMPILER_ID STREQUAL "Intel" AND NOT WIN32)
+    list(APPEND flags_to_test -static-intel -w)
 endif()
 
+test_apply_compiler_flags(flags_to_test CMAKE_C_FLAGS)
```
Comment (1480c1) [inline Source/Lib/Common/ASM_SSE4_1/CMakeLists.txt]:
```suggestion
test_apply_compiler_flags(${flags_to_test})
```

File: Source/Lib/Common/ASM_SSSE3/CMakeLists.txt [hunk 5]
```diff
@@ -14,23 +14,14 @@ include_directories(${PROJECT_SOURCE_DIR}/Source/API/
...
-endforeach()
+set(flags_to_test -mssse3)
 
-if(CMAKE_C_COMPILER_ID STREQUAL "Intel")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
+if(CMAKE_C_COMPILER_ID STREQUAL "Intel" AND NOT WIN32)
+    list(APPEND flags_to_test -static-intel -w)
 endif()
 
+test_apply_compiler_flags(flags_to_test CMAKE_C_FLAGS)
```
Comment (1480c1) [inline Source/Lib/Common/ASM_SSSE3/CMakeLists.txt]:
```suggestion
test_apply_compiler_flags(${flags_to_test})
```

File: Source/Lib/Common/CMakeLists.txt [hunk 6]
```diff
@@ -5,6 +5,17 @@
 
 # Libraries Directory CMakeLists.txt
 
+function(test_apply_compiler_flags flags current_flags)
```
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]:
```suggestion
function(test_apply_compiler_flags flags)
```

File: Source/Lib/Common/CMakeLists.txt (hunk 1 above)
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]:
```suggestion
            set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${cflag}" PARENT_SCOPE)
```

**Comment by DesWurstes** [comment]
Thanks for the review! I've removed the extra parameter, but I disagree with you about replacing

```
test_apply_compiler_flags(flags_to_test)
```
with
```
test_apply_compiler_flags(${flags_to_test})
```

, which causes problems (the foreach doesn't run).

**Comment by 1480c1** [comment]
https://travis-ci.com/1480c1/SVT-AV1/builds/119206431 It successfully ran for me though? Can you paste your cmake error?

File: Source/Lib/Common/CMakeLists.txt [hunk 7]
```diff
@@ -5,6 +5,17 @@
 
 # Libraries Directory CMakeLists.txt
 
+function(test_apply_compiler_flags flags)
```
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]:
```suggestion
function(test_apply_compiler_flags)
```

File: Source/Lib/Common/CMakeLists.txt [hunk 8]
```diff
@@ -5,6 +5,17 @@
...
+    foreach(cflag ${${flags}})
```
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]:
```suggestion
    foreach(cflag ${ARGV})
```

**Comment by 1480c1** [comment]
Actually, it might be better to use a macro instead of a function.

**Comment by DesWurstes** [comment]
Made it a macro. Thanks again for the reviews

**Comment by 1480c1** [comment]
> CMake's counter-intuitive pointer system

Honestly, agreed

**Comment by hassount** [comment]
@kylophone good to go with this ?

**Comment by hassount** [comment]
@DesWurstes @1480c1 
Can we validate that this PR is still good to go on top of the master ?

**Comment by 1480c1** [comment]
Although this may need to change to add avx512

**Comment by 1480c1** [comment]
@DesWurstes, can you rebase and change avx512 too?

**Comment by DesWurstes** [comment]
Should be ready once the tests pass.

**Comment by 1480c1** [comment]
Seems like it's good now

File: Source/Lib/Common/CMakeLists.txt [hunk 9]
```diff
@@ -5,6 +5,17 @@
...
 
+macro(test_apply_compiler_flags)
+    foreach(cflag ${ARGV})
+        string(REGEX REPLACE "[^A-Za-z0-9]" "_" cflag_var "${cflag}")
+        set(test_c_flag "C_FLAG${cflag_var}")
+        check_c_compiler_flag(${cflag} "${test_c_flag}")
+        if(${test_c_flag})
+            set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${cflag}")
+        endif()
+    endforeach()
```
Comment (1480c1) [inline Source/Lib/Common/CMakeLists.txt]:
```suggestion
    endforeach()
    if(CMAKE_BUILD_TYPE MATCHES "[Dd][Ee][Bb][Uu][Gg]")
        message(STATUS "${CMAKE_CURRENT_SOURCE_DIR}: ${CMAKE_C_FLAGS}")
    endif()
```

Comment (1480c1) [reply Source/Lib/Common/CMakeLists.txt]: Other than this, everything is good

**Comment by DesWurstes** [comment]
Done.

**Comment by mafonsoN** [comment]
Looks good to go.

**Pull request merged**